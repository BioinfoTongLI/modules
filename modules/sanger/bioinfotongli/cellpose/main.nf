container_version = "0.1.0"
params.debug = false
params.cellpose_model_dir = "./"

process BIOINFOTONGLI_CELLPOSE {
    tag "${meta.id}"
    debug params.debug
    cache true

    label "gpu"
    label "medium_mem"

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        "quay.io/bioinfotongli/tiled_cellpose:${container_version}":
        "quay.io/bioinfotongli/tiled_cellpose:${container_version}"}"
    containerOptions = {
            workflow.containerEngine == "singularity" ? "--cleanenv --nv -B ${params.cellpose_model_dir}:/tmp/cellpose_models -B ${params.NUMBA_CACHE_DIR}:/tmp/numba_cache":
            ( workflow.containerEngine == "docker" ? "--gpus all -v ${params.cellpose_model_dir}:/tmp/cellpose_models": null )
    }

    publishDir params.out_dir + "/naive_cellpose_segmentation"

    input:
    tuple val(meta), val(x_min), val(y_min), val(x_max), val(y_max), path(image), val(cell_diameter)

    output:
    tuple val(meta), val(cell_diameter), path("${stem}/${stem}_cp_outlines.txt"), emit: outlines, optional: true
    tuple val(meta), val(cell_diameter), path("${stem}/${stem}_cp_outlines.wkt"), emit: wkts
    tuple val(meta), val(cell_diameter), path("${stem}/${stem}*png"), emit: cp_plots, optional: true
    path "versions.yml"           , emit: versions

    script:
    stem = "${meta.id}-${x_min}_${y_min}_${x_max}_${y_max}-diam_${cell_diameter}"
    def args = task.ext.args ?: ''  
    """
    export CELLPOSE_LOCAL_MODELS_PATH=/tmp/cellpose_models
    export NUMBA_CACHE_DIR=/tmp/numba_cache
    /opt/conda/bin/python /scripts/cellpose_seg.py run \
        --image ${image} \
        --x_min ${x_min} \
        --y_min ${y_min} \
        --x_max ${x_max} \
        --y_max ${y_max} \
        --cell_diameter ${cell_diameter} \
        --out_dir "${stem}" \
        ${args}
    
    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        : \$(echo \$(/scripts/cellpose_seg.py version 2>&1) | sed 's/^.*cellpose_seg.py //; s/Using.*\$//' ))
    END_VERSIONS
    """
}
