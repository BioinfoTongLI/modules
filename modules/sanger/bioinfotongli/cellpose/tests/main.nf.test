nextflow_process {

    name "Test Process BIOINFOTONGLI_CELLPOSE"
    script "../main.nf"
    config "./nextflow.config"
    process "BIOINFOTONGLI_CELLPOSE"

    tag "modules"
    tag "modules_sanger"
    tag "bioinfotongli"
    tag "bioinfotongli/cellpose"

    test("mouse_heart - tiff") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test'],
                    0, 0, 100, 100,
                    file(params.modules_testdata_base_path + 'imaging/tiff/mindagap.mouse_heart.wga.tiff', checkIfExists: true),
                    20,
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("mouse_heart - tiff - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test',],
                    0, 0, 100, 100,
                    file(params.modules_testdata_base_path + 'imaging/tiff/mindagap.mouse_heart.wga.tiff', checkIfExists: true),
                    20,
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
