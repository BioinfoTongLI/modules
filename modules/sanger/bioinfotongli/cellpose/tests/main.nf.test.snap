{
    "mouse_heart - tiff - stub": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test-0_0_100_100-diam_20_cp_outlines.wkt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    
                ],
                "3": [
                    "versions.yml:md5,a60a2dc516b21193ec060157456d5d50"
                ],
                "cp_plots": [
                    
                ],
                "outlines": [
                    
                ],
                "versions": [
                    "versions.yml:md5,a60a2dc516b21193ec060157456d5d50"
                ],
                "wkts": [
                    [
                        {
                            "id": "test"
                        },
                        "test-0_0_100_100-diam_20_cp_outlines.wkt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-02-06T22:10:30.731647097"
    },
    "mouse_heart - tiff - 500": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test-0_0_500_500-diam_20_cp_outlines.txt:md5,adb12dc55d990539b15f3ce61d7c8efe"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test-0_0_500_500-diam_20_cp_outlines.wkt:md5,fe837bbb6adc3d13928ed9b3405e4d9c"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test-0_0_500_500-diam_20_cp_masks.png:md5,a1e143c85d8314e9abe9d2d35ea9760a",
                            "test-0_0_500_500-diam_20_outlines.png:md5,36b8f6235c68416d267290547010f453"
                        ]
                    ]
                ],
                "3": [
                    "versions.yml:md5,707d2d840e52d9f1debb52c25faf32fb"
                ],
                "cp_plots": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test-0_0_500_500-diam_20_cp_masks.png:md5,a1e143c85d8314e9abe9d2d35ea9760a",
                            "test-0_0_500_500-diam_20_outlines.png:md5,36b8f6235c68416d267290547010f453"
                        ]
                    ]
                ],
                "outlines": [
                    [
                        {
                            "id": "test"
                        },
                        "test-0_0_500_500-diam_20_cp_outlines.txt:md5,adb12dc55d990539b15f3ce61d7c8efe"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,707d2d840e52d9f1debb52c25faf32fb"
                ],
                "wkts": [
                    [
                        {
                            "id": "test"
                        },
                        "test-0_0_500_500-diam_20_cp_outlines.wkt:md5,fe837bbb6adc3d13928ed9b3405e4d9c"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-02-06T22:09:35.002556103"
    },
    "mouse_heart - tiff - 200": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test-0_0_200_200-diam_20_cp_outlines.txt:md5,c5b8d1848233c7636be9fd1ed6b00010"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test-0_0_200_200-diam_20_cp_outlines.wkt:md5,78ad1ddefc488c798b46ad3c18025912"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test-0_0_200_200-diam_20_cp_masks.png:md5,96c74f7dc06a1ffb1143ad03c4796029",
                            "test-0_0_200_200-diam_20_outlines.png:md5,b3bdb19c06bf039f6c6bfaf051ff32fa"
                        ]
                    ]
                ],
                "3": [
                    "versions.yml:md5,707d2d840e52d9f1debb52c25faf32fb"
                ],
                "cp_plots": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test-0_0_200_200-diam_20_cp_masks.png:md5,96c74f7dc06a1ffb1143ad03c4796029",
                            "test-0_0_200_200-diam_20_outlines.png:md5,b3bdb19c06bf039f6c6bfaf051ff32fa"
                        ]
                    ]
                ],
                "outlines": [
                    [
                        {
                            "id": "test"
                        },
                        "test-0_0_200_200-diam_20_cp_outlines.txt:md5,c5b8d1848233c7636be9fd1ed6b00010"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,707d2d840e52d9f1debb52c25faf32fb"
                ],
                "wkts": [
                    [
                        {
                            "id": "test"
                        },
                        "test-0_0_200_200-diam_20_cp_outlines.wkt:md5,78ad1ddefc488c798b46ad3c18025912"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-02-06T22:09:56.476928706"
    },
    "mouse_heart - tiff - 200 - large cell diameter": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test-0_0_200_200-diam_50_cp_outlines.txt:md5,ddc6f3da007fb90b5c9ea30c9a3ae807"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test-0_0_200_200-diam_50_cp_outlines.wkt:md5,bf5ce6d27e28f6985c4dbdfbf5d3f0ab"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test-0_0_200_200-diam_50_cp_masks.png:md5,2385b7e1d3a3432166029074d43b15dd",
                            "test-0_0_200_200-diam_50_outlines.png:md5,129fd998c787dc5a62575a3c8a7f2bb2"
                        ]
                    ]
                ],
                "3": [
                    "versions.yml:md5,707d2d840e52d9f1debb52c25faf32fb"
                ],
                "cp_plots": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test-0_0_200_200-diam_50_cp_masks.png:md5,2385b7e1d3a3432166029074d43b15dd",
                            "test-0_0_200_200-diam_50_outlines.png:md5,129fd998c787dc5a62575a3c8a7f2bb2"
                        ]
                    ]
                ],
                "outlines": [
                    [
                        {
                            "id": "test"
                        },
                        "test-0_0_200_200-diam_50_cp_outlines.txt:md5,ddc6f3da007fb90b5c9ea30c9a3ae807"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,707d2d840e52d9f1debb52c25faf32fb"
                ],
                "wkts": [
                    [
                        {
                            "id": "test"
                        },
                        "test-0_0_200_200-diam_50_cp_outlines.wkt:md5,bf5ce6d27e28f6985c4dbdfbf5d3f0ab"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.4"
        },
        "timestamp": "2025-02-06T22:10:18.012291506"
    }
}